vllm/executor/ray_gpu_executor.py

```diff
@@ -335,19 +336,35 @@ async def _run_workers_async(
         **kwargs,
     ) -> Any:
         """Runs the given method on all workers."""
-        coros = []
-
         if driver_args is None:
             driver_args = args
         if driver_kwargs is None:
             driver_kwargs = kwargs
 
-        coros.append(
-            self.driver_executor(method, *driver_args, **driver_kwargs))
-
         # Run the ray workers asynchronously.
-        for worker in self.workers:
-            coros.append(worker.execute_method.remote(method, *args, **kwargs))
 
-        all_outputs = await asyncio.gather(*coros)
+        for pp_rank in range(self.parallel_config.pipeline_parallel_size):
+            coros = []
+            # Locks are necessary for correctness in the TP + PP case.
+            async with self.pp_locks[pp_rank]:
+                for tp_rank in range(
+                        self.parallel_config.tensor_parallel_size):
+                    rank = (pp_rank * self.parallel_config.tensor_parallel_size
+                            ) + tp_rank
+                    if rank == 0:
+                        coros.append(
+                            self.driver_executor(method, *driver_args,
+                                                 **driver_kwargs))
+                    else:
+                        worker = self.workers[rank - 1]
+                        if tp_rank == 0:
+                            coros.append(
+                                worker.execute_method.remote(
+                                    method, *driver_args, **driver_kwargs))
+                        else:
+                            coros.append(
+                                worker.execute_method.remote(
+                                    method, *args, **kwargs))
+                all_outputs = await asyncio.gather(*coros)
```

Comment:
I am a bit confused by this loop. Please help with my understanding, thanks! @andoorve 

Say if we have `pp_size = 2` and `tp_size=1`, we'll iterate over the two ranks from 0 to 1 (outer loop, inner loop diminishes).

On pp_rank=0, we will launch the execution of the first stage on its corresponding GPU (L355) and await its completion (L368). Once L368 returns, we will proceed to rank 1. The part I am confused, if in rank=0's execution, we launch a NCCL `send`, this `send` won't return unless we have rank=1 to launch its corresponding `recv`, right? In this case, that L368 `await` will never return hence block the loop to launch its corresponding `recv`?

Please correct if I am wrong, appreciate your help! @andoorve !

Reply:
followup: 

I debugged a bit and did a simple experiment:

I file a single request. 
- On rank=0, I do not change any model code and just launch the NCCL send as your code did
- while on rank=1, I changed your code and do not launch a `recv` ( but using fake values of hidden_states). 

I added a few prints in your code and found that the send on rank=0 can still passes! Given that your `send_to_next_rank` and `recv_from_prev_rank` are indeed using the synchronous  send/recv API, this is extremely suspicious  that the current code might have messed up some send/recv pairings...

Reply:
Hi @zhisbug

Thanks for your thorough investigation! I was also initially puzzled by this and did not expect this code to work - send/recv should be blocking and thus we should hang while waiting at L368 `await`. At the time, I thought this might be due to a ray quirk - i.e. returning early somehow.

At the time I dismissed it because:

a) The output we get is correct
b) When I printed the sent/recv'd tensors those appeared to be correct
c) When I check the trace with `nsys` I see `send` and `recv` matching up there.

However, what you are saying is true - I do see the `send` ending before the `recv` begins when I print timestamps. It needs a more in-depth look.

I tried your debugging method but I do see hangs when I do it this way which we would expect:

> - On rank=0, I do not change any model code and just launch the NCCL send as your code did
> - while on rank=1, I changed your code and do not launch a recv ( but using fake values of hidden_states).

May I know your exact modifications?

Reply:
if you change your llama forward function to the following, in which I let the first `profile_run` to complete a full send/recv by making a special case based on shape (in fact, I also found the profile run won't go through your raygpuexecutor code), but the following send/recv to be only partial -- only send, no recv.
```
    def forward(
        self,
        input_ids: Optional[torch.Tensor],
        positions: torch.Tensor,
        kv_caches: List[torch.Tensor],
        attn_metadata: AttentionMetadata,
        inputs_embeds: Optional[torch.Tensor] = None,
    ) -> torch.Tensor:
        if is_pipeline_model_parallel_first_rank():
            if inputs_embeds is not None:
                hidden_states = inputs_embeds
            else:
                hidden_states = self.get_input_embeddings(input_ids)
            residual = None
        else:
            if inputs_embeds is not None:
                sizes = list(inputs_embeds.size())
            else:
                sizes = list(input_ids.size()) + [self.config.hidden_size]
            print(f"{sizes}")
            if sizes[0] == 2048:
                hidden_states, residual = recv_prev_rank(
                    2, torch.Size(sizes), self.embed_tokens.weight.dtype,
                    self.embed_tokens.weight.device)
            else:
                if inputs_embeds is not None:
                    hidden_states = inputs_embeds
                else:
                    hidden_states = self.get_input_embeddings(input_ids)
                residual = None

        for i in range(self.start_layer, self.end_layer):
            layer = self.layers[i]
            hidden_states, residual = layer(
                positions,
                hidden_states,
                kv_caches[i - self.start_layer],
                attn_metadata,
                residual,
            )
```

Then you can launch the openai server and send one prompt using the following two commands:
```
CUDA_LAUNCH_BLOCKING=1 python -m vllm.entrypoints.openai.api_server --model JackFram/llama-160m  --swap-space 16 --pipeline-parallel-size 2 --enforce-eager
```

```
python benchmark_serving.py --backend vllm --model JackFram/llama-160m --dataset-name sharegpt --dataset-path ~/sharegpt.json --num-prompts 1 --sharegpt-output-len 5
```


If you add a few prints at the end of the send API:

```
def send_next_rank(tensors: List[torch.Tensor]) -> None:
    """Send the tensors to the next pipeline model parallel rank."""
    print(f"global rank {torch.distributed.get_rank()} sending...")
    combined_tensor = torch.cat(tensors, dim=0)
    torch.distributed.send(combined_tensor,
                           get_pipeline_model_parallel_next_rank(),
                           get_pipeline_model_parallel_group())
    print(f"global rank {torch.distributed.get_rank()} sent done")
```
You will observe the first `send` passes even I do not launch recv for it! After the first send/recv, the server will hang at the line which I print "global rank {torch.distributed.get_rank()} sending..." . 

This is extremely strange because we do not launch a coressponding recv for the first send, how could?

Reply:
Ok, I think I have a good idea of what is going on here, and why what we are still seeing the correct output even though `send` and `recv` here are not operating as we expect - but it's harder to figure out why that's happening.

My hypothesis is that `torch.distributed.send` is enqueuing the correct send operation on the relevant CUDA stream. However, instead of blocking waiting for that NCCL operation to complete, it is instead returning immediately from the Python function. I.e. it is operating similar to the non-blocking `isend` function which is not what we expect.

This is why you see the first send "go-through" (It is not actually sending just returning from the Python function) and blocking on the second send.

I modified `communication_op.py` to include some `print` statements and `time.sleep`s.

```
def send_next_rank(tensors: List[torch.Tensor]) -> None:
    """Send the tensors to the next pipeline model parallel rank."""
    combined_tensor = torch.cat(tensors, dim=0)
    torch.cat(tensors, dim=0)
    print (f'SEND STARTING {time.time()}', flush=True)
    torch.distributed.send(combined_tensor,
                           get_pipeline_model_parallel_next_rank(),
                           get_pipeline_model_parallel_group())
    print(f'SEND SUM: {combined_tensor.sum()}', flush=True)
    print (f'SEND COMPLETED {time.time()}', flush=True)
    time.sleep(5)


def recv_prev_rank(num_tensors: int, sizes: torch.Size, dtype: torch.dtype,
                   device: torch.device) -> List[torch.Tensor]:
    sizes = list(sizes)
    """Receive tensors from the previous pipeline model parallel rank."""
    combined_tensor = torch.empty([sizes[0] * num_tensors] + sizes[1:],
                                  dtype=dtype,
                                  device=device)
    time.sleep(5)
    print (f'RECV STARTING {time.time()}', flush=True)
    torch.distributed.recv(combined_tensor,
                           get_pipeline_model_parallel_prev_rank(),
                           get_pipeline_model_parallel_group())
    print(f'RECV SUM: {combined_tensor.sum()}', flush=True)
    print (f'RECV COMPLETED {time.time()}', flush=True)
    return torch.chunk(combined_tensor, num_tensors, dim=0)
```

This gives the following output (I also include print statements for entering and exiting an outer loop iteration in `ray_gpu_executor.py`.

```
PP RANK 0 STARTED! at 1715983628.996715
SEND STARTING 1715983629.1642857
SEND SUM: -248.0
SEND COMPLETED 1715983629.1664162
PP RANK 0 DONE! at 1715983634.1697052
PP RANK 1 STARTED! at 1715983634.169776
(RayWorkerWrapper pid=2335407) RECV STARTING 1715983639.196885
(RayWorkerWrapper pid=2335407) RECV SUM: -248.0
(RayWorkerWrapper pid=2335407) RECV COMPLETED 1715983639.1983254
PP RANK 1 DONE! at 1715983639.2434692
```

Here, the time in the python portion of the `send` function is completely disjoint from the time spent in the `recv` portion (and PP rank 0 is done before the subsequent RECV is started). However, the checksums are consistent. This is consistent with what you saw in your experiment where the second `send` was blocked by the first (since there was no matching `recv` for the first `send`) even though the Python function completed.

I tried to gather `nsys` trace here for one more piece of evidence, but unfortunately was not able to get both the GPU traces for some reason.

Of course, although in this case this behaviour seems to be working in our favour it is not at all consistent with what we expect of `torch.distributed.send` and `torch.distributed.recv` which is to block the Python thread.

Reply:
To continue, I also tried a small unit test and here the semantics are respected. That is, `send` blocks its process until the `recv` is completed.

```
import time
import os
import torch
import torch.distributed as dist
import torch.multiprocessing as mp

def run(rank, size):
    """ Distributed function to be implemented later. """
    device = torch.device(f"cuda:{rank}")
    if rank == 0:
        tens = torch.ones([33, 4096], dtype=torch.bfloat16, device=device)
        dist.send(tens, dst=1)
        print(f'SEND COMPLETED {time.ctime()}', flush=True)
    else:
        tens = torch.empty([33, 4096], dtype=torch.bfloat16, device=device)
        time.sleep(10)
        print(f'ABOUT TO BEGIN RECV {time.ctime()}', flush=True)
        dist.recv(tens, src=0)
        print (f'{tens}', flush=True)

def init_process(rank, size, fn, backend='nccl'):
    """ Initialize the distributed environment. """
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = '29501'
    dist.init_process_group(backend, rank=rank, world_size=size)
    fn(rank, size)


if __name__ == "__main__":
    size = 2
    processes = []
    mp.set_start_method("spawn")
    for rank in range(size):
        p = mp.Process(target=init_process, args=(rank, size, run, 'nccl'))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```

I have been trying to see if there are some env variables that perhaps Ray or vLLM might have set to change PyTorch's behaviour but not successful so far.